src/V3CUse.cpp

```diff
@@ -46,52 +48,34 @@ class CUseVisitor final : public VNVisitor {
 
     // MEMBERS
     AstNodeModule* const m_modp;  // Current module
-    std::set<std::pair<VUseType, std::string>> m_didUse;  // What we already used
-    bool m_dtypesImplOnly = false;
+    std::map<std::string, std::pair<FileLine*, uint8_t>> m_didUse;  // What we already used
 
     // METHODS
     void addNewUse(AstNode* nodep, VUseType useType, const string& name) {
-        if (m_dtypesImplOnly
-            && (useType == VUseType::INT_INCLUDE || useType == VUseType::INT_FWD_CLASS))
-            return;
-
-        if (m_didUse.emplace(useType, name).second) {
-            AstCUse* const newp = new AstCUse{nodep->fileline(), useType, name};
-            m_modp->addStmtsp(newp);
-            UINFO(8, "Insert " << newp << endl);
+        auto e = m_didUse.emplace(name, std::make_pair(nodep->fileline(), useType));
+        if (e.second || ((e.first->second.second & useType) != useType)) {
+            e.first->second.second |= useType;
         }
     }
 
     // VISITORS
     void visit(AstClassRefDType* nodep) override {
-        if (nodep->user1SetOnce()) return;  // Process once
         addNewUse(nodep, VUseType::INT_FWD_CLASS, nodep->classp()->name());
     }
     void visit(AstCFunc* nodep) override {
-        if (nodep->user1SetOnce()) return;  // Process once
+        if (nodep->user1SetOnce()) return;
         iterateAndNextNull(nodep->argsp());
-
-        {
-            VL_RESTORER(m_dtypesImplOnly);
-            m_dtypesImplOnly = true;
-
-            iterateAndNextNull(nodep->initsp());
-            iterateAndNextNull(nodep->stmtsp());
-            iterateAndNextNull(nodep->finalsp());
-        }
+        if (nodep->stmtsp())
+            nodep->stmtsp()->forall([this](AstCReturn* retp) -> bool {
+                visit(retp);
+                return true;
+            });
```

Comment:
This is much more expensive than the old way, and also a trap if we add code that expects visitors to work as they usually do.  So I'd prefer to keep the iterates.

Reply:
Is `forall` generally more expensive or is it simply because we might take shortcuts in visitors and skip some visits? THis looked to me linke a good place to use `forall` as the regular behaviour of CUse visitor should be discarded when visiting implementations as CUses are irrelevant to implementation and we only care about returns.

Reply:
FWIW `forall` (and `exists`) shortcuts if it can. I.e.: `forall` will stop as soon as it encounters a node that fails the predicate, and `exists` will stop as soon as it encounter a node that passes the predicate.